client/gateway_client.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// GetActiveSessions retrieves active sessions for a list of app addresses and a service ID.
+// It verifies that each app delegates to the gateway and is staked for the requested service.
+// This method encapsulates the shared logic between centralized and delegated gateway modes.
+func (c *GatewayClient) GetActiveSessions(
+	ctx context.Context,
+	serviceID sdk.ServiceID,
+	appAddresses []string,
+) ([]sessiontypes.Session, error) {
+	if len(appAddresses) == 0 {
+		return nil, fmt.Errorf("no app addresses provided for service %s", serviceID)
+	}
+
+	var activeSessions []sessiontypes.Session
+
+	for _, appAddr := range appAddresses {
+		// Retrieve the session for the app
+		session, err := c.GetSession(ctx, serviceID, appAddr)
+		if err != nil {
+			return nil, fmt.Errorf("%w: app: %s, error: %w",
+				ErrProtocolContextSetupAppFetchErr,
+				appAddr,
+				err,
+			)
+		}
+
+		app := session.Application
+
+		// Verify the app is staked for the requested service
+		if !appIsStakedForService(serviceID, app) {
+			return nil, fmt.Errorf("%w: app: %s",
+				ErrProtocolContextSetupAppNotStaked,
+				app.Address,
+			)
+		}
+
+		// Verify the app delegates to the gateway
+		if !c.gatewayHasDelegationForApp(app) {
+			return nil, fmt.Errorf("%w: gateway: %s, app: %s",
+				ErrProtocolContextSetupAppDelegation,
+				c.gatewayAddress,
+				app.Address,
+			)
+		}
+
+		activeSessions = append(activeSessions, session)
+	}
+
+	if len(activeSessions) == 0 {
+		return nil, fmt.Errorf("%w: service %s",
+			ErrProtocolContextSetupNoSessions,
+			serviceID,
+		)
+	}
+
+	return activeSessions, nil
+}
+
+// gatewayHasDelegationForApp checks if the gateway has delegation for the application.
+func (c *GatewayClient) gatewayHasDelegationForApp(app *apptypes.Application) bool {
+	return slices.Contains(app.DelegateeGatewayAddresses, c.gatewayAddress)
+}
+
+// appIsStakedForService checks if the application is staked for the specified service.
+func appIsStakedForService(serviceID sdk.ServiceID, app *apptypes.Application) bool {
+	for _, serviceConfig := range app.GetServiceConfigs() {
+		if serviceConfig.GetServiceId() == string(serviceID) {
+			return true
+		}
+	}
+	return false
+}
+
+// ValidateRelayResponse validates the RelayResponse and verifies the supplier's signature.
+//
+// - Returns the RelayResponse, even if basic validation fails (may contain error reason).
+// - Verifies supplier's signature with the provided publicKeyFetcher.
+func (c *GatewayClient) ValidateRelayResponse(
+	ctx context.Context,
+	supplierAddress sdk.SupplierAddress,
+	relayResponseBz []byte,
+) (*servicetypes.RelayResponse, error) {
+	relayResponse := &servicetypes.RelayResponse{}
+	if err := relayResponse.Unmarshal(relayResponseBz); err != nil {
+		return nil, err
+	}
+
+	if err := relayResponse.ValidateBasic(); err != nil {
+		// Even if the relay response is invalid, return it (may contain failure reason)
+		return relayResponse, err
```

Comment:
Please define custom errors for these: the current version of SDK already has that.

Reply:
I've updated the gateway client file to greatly improve custom error definitions and logging of said errors.